python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll

```diff
@@ -9,3 +9,136 @@ private import semmle.python.dataflow.new.TaintTracking
 private import semmle.python.dataflow.new.RemoteFlowSources
 private import experimental.semmle.python.Concepts
 private import semmle.python.ApiGraphs
+
+private module NoSQL {
+  /** Gets a reference to a `MongoClient` DB. */
+  private API::Node mongoClientInstance() {
+    result = API::moduleImport("pymongo").getMember("MongoClient").getReturn() or
+    result =
+      API::moduleImport("flask_mongoengine")
+          .getMember("MongoEngine")
+          .getReturn()
+          .getMember("get_db")
+          .getReturn() or
+    result = API::moduleImport("mongoengine").getMember("connect").getReturn() or
+    result = API::moduleImport("flask_pymongo").getMember("PyMongo").getReturn()
+  }
+
+  /** Gets a reference to a `MongoClient` DB. */
+  private DataFlow::LocalSourceNode mongoClientDB(DataFlow::TypeTracker t) {
+    t.start() and
+    (
+      exists(SubscriptNode subscript | result.asCfgNode() = subscript |
+        subscript.getObject() = mongoClientInstance().getAUse().asCfgNode()
+      )
+      or
+      result.(DataFlow::AttrRead).getObject() = mongoClientInstance().getAUse()
+    )
+    or
+    exists(DataFlow::TypeTracker t2 | result = mongoClientDB(t2).track(t2, t))
+  }
+
+  /** Gets a reference to a `MongoClient` DB. */
+  private DataFlow::Node mongoClientDB() {
+    mongoClientDB(DataFlow::TypeTracker::end()).flowsTo(result)
+  }
+
+  /** Gets a reference to a `MongoClient` collection. */
+  private DataFlow::LocalSourceNode mongoClientCollection(DataFlow::TypeTracker t) {
+    t.start() and
+    (
+      exists(SubscriptNode subscript | result.asCfgNode() = subscript |
+        subscript.getObject() = mongoClientDB().asCfgNode()
+      )
+      or
+      result.(DataFlow::AttrRead).getObject() = mongoClientDB()
+    )
+    or
+    exists(DataFlow::TypeTracker t2 | result = mongoClientCollection(t2).track(t2, t))
+  }
+
+  /** Gets a reference to a `MongoClient` collection. */
+  private DataFlow::Node mongoClientCollection() {
+    mongoClientCollection(DataFlow::TypeTracker::end()).flowsTo(result)
+  }
+
+  /** This class represents names of find_* relevant MongoClient collection level operation methods. */
+  private class MongoClientMethodNames extends string {
+    MongoClientMethodNames() {
+      // the find_one_or_404 method is only found in the Pymongo Flask library.  
```

Comment:
I'm not really sure how I should have treated this. `find_one_or_404` is a method exclusive to the Pymongo Flask library. Under the hood, Pymongo Flask will capture any calls made to MongoClient when `find_one_or_404` is called on a MongoClient instance. 

I checked and no other library has this method defined. Because of that, I thought it would be ok to lump it into this class. I don't see how any other developer could use this method in another library since it's defined and I thought that trying to distinguish this one single method from all of the base-class MongoClient sinks within the CodeQL query would be more trouble than it's worth.

Welcome any input on this.

Reply:
With the current setup of the code in this PR, I think this is an OK solution :+1:

Reply:
Resolving this as no other action is required.